array.rb

```diff
@@ -1,4 +1,16 @@
 class Array
+  def each
+    unless block_given?
+      return to_enum(:each) { self.length }
+    end
+    i = 0
+    while i < self.length
+      yield self[i]
+      i = i.succ
```

Comment:
Why use `i.succ` over `i += 1`? You may need to guard against people redefining `succ` as well.

Reply:
> Why use i.succ over i += 1?

Faster for the interpreter. `putobject 1` `opt_plus` vs `opt_succ`.

> You may need to guard against people redefining succ as well.

The discussion applies whether it's `+` or `succ`. We'd use `Primitive` if we want to guard that, but I'm not sure if we really need to use `Primitive` for every single method call that was not a method call before.

Reply:
> Faster for the interpreter. putobject 1 opt_plus vs opt_succ. Not a problem for YJIT, so I'm open to using i += 1 though.

I don't have an opinion on this. I was just curious. Thanks for the explanation.

> The discussion applies whether it's + or succ.

Right. The method redefinition is a general problem. I didn't mean to imply that it only applies to `succ`. I don't disagree that redefining such methods is bound to break something. I'm just noting that it is a behavioral difference between the Ruby and C implementations.

```ruby
class Fixnum
  def succ
    3
  end
end

[1, 2, 3, 4].each { |x| puts x }
```

I guess this hits on a larger discussion about what can we assume doesn't change. We've run into similar issues with TruffleRuby and in some cases we've made similar judgment calls.